symcrypt-sys/build.rs

```diff
@@ -0,0 +1,90 @@
+extern crate bindgen;
+
+use std::env;
+use std::path::PathBuf;
+
+fn main() {
+    #[cfg(target_os = "windows")] 
+    {
+        println!("cargo:rustc-link-search=native=C:/Windows/System32/"); // ! Work around, looking for better solution
+        println!("cargo:libdir=./inc");
+        println!("cargo:rustc-link-lib=dylib=symcrypttestmodule"); // test module used in lieu of official symcrypt dll
+        // this dll will be in Windows/System32. This is to mirror future plans; as symcrypt is planned to ship with windows
+        // symcrypttestmodule* files must be placed in Windows/System32/ as a workaround at the moment.
+    }
+    
+    #[cfg(target_os = "linux")]
+    {   
+        println!("cargo:libdir=./inc");
+        println!("cargo:rustc-link-lib=dylib=symcrypt"); // the lib prefix for libsymcrypt is implied on linux
+
+        // TODO: Create a script that copies all libsymcrypt.so* files from SymCrypt path to /lib/x86_64-linux-gnu/
+        // The ld linker will look for the symcrypt.so files within /lib/x86_64-linux-gnu/. No need to set a hardcoded path.
+    }
+
+    // TODO: Factor out binding generation. Bindgen should only be run manually with updates to underlying SymCrypt code that 
+    // we decide what to take.
+    println!("cargo:rerun-if-changed=inc/wrapper.h");
+
+    // TODO: Discuss if factoring the .allowlist_functions to another file is better approach
+
+    let bindings = bindgen::Builder::default()
+        .header("inc/wrapper.h")
+        .clang_arg("-v")
+        .parse_callbacks(Box::new(bindgen::CargoCallbacks))
+        // ALLOWLIST 
+
+        // INIT FUNCTIONS
+        .allowlist_function("SymCryptInit")
+
+        // HASH FUNCTIONS 
+        .allowlist_function("^(SymCryptSha256.*)$")
+        .allowlist_function("^(SymCryptSha384.*)$")
+        .allowlist_var("SYMCRYPT_SHA256_RESULT_SIZE")
+        .allowlist_var("SYMCRYPT_SHA384_RESULT_SIZE")
+
+        // HMAC FUNCTIONS
+        .allowlist_function("^(SymCryptHmacSha256.*)$")
+        .allowlist_function("^(SymCryptHmacSha384.*)$")
+
+        // GCM FUNCTIONS
+        .allowlist_function("^(SymCryptGcm.*)$")
+
+        .allowlist_function("SymCryptChaCha20Poly1305Encrypt")
+        .allowlist_function("SymCryptChaCha20Poly1305Decrypt")
+
+        .allowlist_function("SymCryptTlsPrf1_2ExpandKey")
+        .allowlist_function("SymCryptTlsPrf1_2Derive")
+        .allowlist_function("SymCryptTlsPrf1_2")
+
+        // HKDF functions 
+        .allowlist_function("^(SymCryptHkdf.*)$")
+
+        // ECDH Key Agreement
+        .allowlist_var("SymCryptEcurveParamsNistP256")
+        .allowlist_var("SymCryptEcurveParamsNistP384")

```

Comment:
Were you able to resolve the linking issue with these exported constants?

Reply:
Yes, we were able to fix them with this:

`#[link(name = "symcrypttestmodule", kind = "dylib")]
extern "C" {
    pub static SymCryptSha256Algorithm: PCSYMCRYPT_HASH;
}`

Copied from internal email thread:
"
The core issue is something mentioned at the bottom of the docs for [EXPORTS | Microsoft Learn](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Flearn.microsoft.com%2Fen-us%2Fcpp%2Fbuild%2Freference%2Fexports%3Fview%3Dmsvc-170&data=05%7C01%7Ckhangnguyen%40microsoft.com%7Cc4649113c171424fe3c108db685f3dbb%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638218530497811201%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=UCMGw0CzGkk1T1TSif2g2tZI7yHaiNXjEjIAxxNyFbk%3D&reserved=0):
> When you export a variable from a DLL by using a .DEF file, you do not have to specify __declspec(dllexport) on the variable. However, in any file that uses the DLL, you must still use __declspec(dllimport) on the declaration of data.
(emphasis mine)
"